### PR TITLE
handle case where user-email not set

### LINKED ISSUE #1706: git_sitrep() vs missing user.email
I cleared out my R/RStudio/Git setup to try to see what new users see, to help R4DS users get up to speed. I accidentally skipped the `usethis::use_git_config(user.name = "Jane Doe", user.email = "jane@example.org")` step, and was somewhat surprised to find that `git_sitrep()` failed rather than telling me about it. It's POSSIBLE this happened due to something being half-set somewhere (it's surprisingly hard to clear out every setting for everything everywhere), but I'm PRETTY sure this could be common.

This causes an error in `pat_sitrep()` in the [emails section](https://github.com/r-lib/usethis/blob/776237c8ab4bf69f7c33d07fc7b0798521316398/R/github_token.R#L184), because `de_facto_email` is NULL.

Maybe add something toward the beginning of `git_sitrep()` to error with a message telling the user to run `use_git_config()`?

``` r
dat <- gert::git_config_global()
if ("user.email" %in% dat$name) {
  old_email <- dat[dat$name == "user.email",]$value
  usethis::use_git_config(user.email = NULL)
}
usethis::git_sitrep()
#> Git config (global)
#> • Name: 'Jon Harmon'
#> • Email: <unset>
#> • Global (user-level) gitignore file: 'C:/Users/jonth/.gitignore'
#> • Vaccinated: TRUE
#> ℹ Defaulting to 'https' Git protocol
#> • Default Git protocol: 'https'
#> • Default initial branch name: 'main'
#> GitHub
#> • Default GitHub host: 'https://github.com'
#> • Personal access token for 'https://github.com': '<discovered>'
#> • GitHub user: 'jonthegeek'
#> • Token scopes: 'gist, repo, user, workflow'
#> • Email(s): 'jonthegeek@gmail.com (primary)', '33983824+jonthegeek@users.noreply.github.com', 'jonthegeek42@gmail.com'
#> Error in grepl(de_facto_email, addresses): invalid 'pattern' argument

if ("user.email" %in% dat$name) {
  usethis::use_git_config(user.email = old_email)
}
```

<sup>Created on 2022-11-30 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

### PR DESCRIPTION
**Pull request by ijlyttle**
fix #1706

While `git_sitrep()` was still loaded into my brain, I thought to take a swing at this. Given that `git_sitrep()` can be run in bits-and-pieces, I thought the feedback should be proximate to where the error was happening.

Following @jonthegeek's reprex:

``` r
dat <- gert::git_config_global()
if ("user.email" %in% dat$name) {
  old_email <- dat[dat$name == "user.email",]$value
  usethis::use_git_config(user.email = NULL)
}
usethis::git_sitrep(scope = "user")
#> 
#> ── Git global (user)
#> • Name: 'Ian Lyttle'
#> • Email: <unset>
#> • Global (user-level) gitignore file: '~/.gitignore'
#> • Vaccinated: TRUE
#> ℹ Defaulting to 'https' Git protocol
#> • Default Git protocol: 'https'
#> • Default initial branch name: 'main'
#> 
#> 
#> ── GitHub user 
#> 
#> • Default GitHub host: 'https://github.com'
#> • Personal access token for 'https://github.com': '<discovered>'
#> • GitHub user: 'ijlyttle'
#> • Token scopes: 'gist, repo, user, workflow'
#> • Email(s): 'ijlyttle@me.com (primary)', 'ijlyttle@users.noreply.github.com', 'ian.lyttle@<work>.com'
#> ✖ Git user's email is not set. Configure using `use_git_config(user.email = "<your email>")`.

if ("user.email" %in% dat$name) {
  usethis::use_git_config(user.email = old_email)
}
```

<sup>Created on 2023-01-11 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup> 

I thought to combine the news item with the other work, but that may be the wrong choice.

Thanks!

### CODE REVIEW

**Comment by jonthegeek** [comment]
Excellent, thanks!

**Comment by jennybc** [comment]
I would expect to see some changes here:

https://github.com/r-lib/usethis/blob/cebe570f2988999fb1c8b0d23290bf1e9faabb52/R/git.R#L353

It's conceivable that the answer is to create a specialized sitrep helper just for the git user, sort of like for the PAT.

I also think we'd want to raise the alarm if the git user's name isn't configured, not just email.

**Comment by ijlyttle** [comment]
Here's another swing at this:

- Whenever the git user is needed, when checking `"git"` and `"github"`, it will complain if `user.name` or `user.email` is not set.
- The git name and email are always reported for `"project"` (new), and noted if there is a local configuration (as before).

In short, it's a bit more verbose and surly.

Nuts and bolts:

- new internal function `git_user_sitrep()`, with helpers `git_user_get()`, `git_user_check()`.
- new internal helper `where_from_scope()`, which could be used if `pat_sitrep()` were to gain a scope argument.

New reprex, including name:

``` r
dat <- gert::git_config_global()
if ("user.name" %in% dat$name) {
  old_name <- dat[dat$name == "user.name",]$value
  usethis::use_git_config(user.name = NULL)
}
if ("user.email" %in% dat$name) {
  old_email <- dat[dat$name == "user.email",]$value
  usethis::use_git_config(user.email = NULL)
}
usethis::git_sitrep(scope = "user")
#> 
#> ── Git global (user)
#> • Name: <unset>
#> • Email: <unset>
#> ✖ Git user's name and email are not set. Configure using `use_git_config(user.name = "<your name>", user.email = "<your email>")`.
#> • Global (user-level) gitignore file: '~/.gitignore'
#> • Vaccinated: TRUE
#> ℹ Defaulting to 'https' Git protocol
#> • Default Git protocol: 'https'
#> • Default initial branch name: 'main'
#> 
#> 
#> ── GitHub user 
#> 
#> • Default GitHub host: 'https://github.com'
#> • Personal access token for 'https://github.com': '<discovered>'
#> • GitHub user: 'ijlyttle'
#> • Token scopes: 'gist, repo, user, workflow'
#> • Email(s): 'ijlyttle@me.com (primary)', 'ijlyttle@users.noreply.github.com', 'ian.lyttle@<work>.com'
#> ✖ Git user's name and email are not set. Configure using `use_git_config(user.name = "<your name>", user.email = "<your email>")`.
if ("user.name" %in% dat$name) {
  usethis::use_git_config(user.name = old_name)
}
if ("user.email" %in% dat$name) {
  usethis::use_git_config(user.email = old_email)
}
```

<sup>Created on 2023-01-18 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

File: R/git.R [hunk 1]
```diff
@@ -446,6 +432,47 @@ git_sitrep <- function(tool = c("git", "github"),
   invisible()
 }
 
+git_user_sitrep <- function(scope = c("user", "project")) {
+  scope <- match.arg(scope)
```
[the remark this answers is not in the document]
Comment (ijlyttle) [reply R/git.R]: updated in f92a930

**Comment by ijlyttle** [comment]
@jennybc Sorry, I got myself turned around: this has `where_from_scope()`, so this will need to go first.

Everything ready on this PR, thanks!

**Comment by jennybc** [comment]
Thanks!

**Pull request merged**